### ISSUE TITLE
Memory leaks in Golang slices

### ISSUE THREAD

**Issue by Rikanishu** [swig/swig#2292]
Hi there!

We had faced with one problem recently when we wrote some code with annoy ANN library (https://github.com/spotify/annoy).
Originally, it's written in C++, but has bindings for Python out of the box. Also, according a separate documentation we can generate a binding layer for Golang. 

The documentation for Golang: https://github.com/spotify/annoy/blob/master/README_GO.rst

The documentation of annoy recommends to use SWIG to generate that layer. And if we try to check the code after that (via tests or with single checking), everything will work fine. But there is a very unpleasant moment: it has memory leaks. And the documentation even warns us that it may take place. And it happens constantly and it reproduces each time as well.

**Repo with generated code**
To reproduce this leaks scenario I've created the repo https://github.com/Rikanishu/annoy-go.
Reproduced with SWIG v4.0.2 / macOS Catalina v10.15.7.

**Overview of the code**
Here is a short overview which code will be generated when we use SWIG.
I selected only needed functions to demonstrate the problem.

The generated function calls a wrapper:

```go
func (p SwigcptrAnnoyIndexAngular) GetNnsByVector(a ...interface{}) {
	argc := len(a)
	if argc == 4 {
		p.GetNnsByVector__SWIG_1(a[0].([]float32), a[1].(int), a[2].(int), a[3].(*[]int))
		return
	}
	if argc == 5 {
		p.GetNnsByVector__SWIG_0(a[0].([]float32), a[1].(int), a[2].(int), a[3].(*[]int), a[4].(*[]float32))
		return
	}
	panic("No match for overloaded function call")
}
```

And finally we fall into this C++ function. I've taken one of them.
https://github.com/Rikanishu/annoy-go/blob/0b0c91e3c22677172147838828af6eb0bb64fbab/annoygomodule_wrap.cxx#L964

```go

void _wrap_AnnoyIndexAngular_getNnsByVector__SWIG_0_annoyindex_2832eef9f6555f92(GoAnnoy::AnnoyIndexAngular *_swig_go_0, _goslice_ _swig_go_1, intgo _swig_go_2, intgo _swig_go_3, _goslice_* _swig_go_4, _goslice_* _swig_go_5) {
  GoAnnoy::AnnoyIndexAngular *arg1 = (GoAnnoy::AnnoyIndexAngular *) 0 ;
  float *arg2 = (float *) 0 ;
  int arg3 ;
  int arg4 ;
  vector< int32_t > *arg5 = (vector< int32_t > *) 0 ;
  vector< float > *arg6 = (vector< float > *) 0 ;
  
  arg1 = *(GoAnnoy::AnnoyIndexAngular **)&_swig_go_0; 
  
  float *v;
  vector<float> w;
  v = (float *)_swig_go_1.array;
  for (int i = 0; i < _swig_go_1.len; i++) {
    w.push_back(v[i]);
  }
  arg2 = &w[0];
  
  arg3 = (int)_swig_go_2; 
  arg4 = (int)_swig_go_3; 
  
  arg5 = new vector<int32_t>();
  
  
  arg6 = new vector<float>();
  
  
  GoAnnoy::AnnoyIndex *swig_b0 = (GoAnnoy::AnnoyIndex *)arg1;
  (swig_b0)->getNnsByVector((float const *)arg2,arg3,arg4,arg5,arg6);
  
  
  {
    _swig_go_4->len = arg5->size();
    _swig_go_4->cap = arg5->size();
    _swig_go_4->array = malloc(_swig_go_4->len * sizeof(intgo));
    for (int i = 0; i < arg5->size(); i++) {
      ((intgo *)_swig_go_4->array)[i] = (intgo)(*arg5)[i];
    }
  }
  
  
  {
    _swig_go_5->len = arg6->size();
    _swig_go_5->cap = arg6->size();
    _swig_go_5->array = malloc(_swig_go_5->len * sizeof(float));
    for (int i = 0; i < arg6->size(); i++) {
      ((float *)_swig_go_5->array)[i] = (float)(*arg6)[i];
    }
  }
  
  
  delete arg5;
  
  
  delete arg6;
  
}

```

And we call this method that way, according the example:

```go
var result []int
t.GetNnsByItem(0, 1000, -1, &result)
fmt.Printf("%v\n", result)
```

And if we do it in loop, we can notice that our memory is running out slowly.

All the magic happen here:
```go
_swig_go_4->array = malloc(_swig_go_4->len * sizeof(intgo));
...
_swig_go_5->array = malloc(_swig_go_5->len * sizeof(float));
```

These arrays won't be reachable for Golang GC.

**My hypothesis why the leak takes place**
The problem is the fact that go allocation of a slice is not the same as the allocation of a similar structure (_goslice_ in SWIG) in c++.
The allocator of Golang always assigns a special mark to any structure that has inner non-nil pointer inside. And this mark will be used by the garbage collector to continue scanning inside of a structure it looks at. And normally each allocation through golang allocation set this mark. Otherwise, if you haven't set it, GC will just interpret this value in memory like a regular scalar, not as a pointer.

> We needed to have some information about the objects since we didn’t have headers. Mark bits are kept on the side and used for marking as well as allocation. Each word has 2 bits associated with it to tell you if it was a scalar or a pointer inside that word. It also encoded whether there were more pointers in the object so we could stop scanning objects sooner than later. We also had an extra bit encoding that we could use as an extra mark bit or to do other debugging things. This was really valuable for getting this stuff running and finding bugs.

[Source](https://go.dev/blog/ismmkeynote)

**Conclusion**
So I suppose that any slice -> vector conversion in SWIG will have the same issue. As I understand, the problem is general for any c++ code that writes to slices from Golang.

**PS.**
To avoid these leaks we've fixed passing of slices for our cases. As a drawback it means all the future updates of this library should be handled manually. And of course it doesn't resolve the problem in general.

Thanks.

**Comment by ojwb** [comment]
@ianlancetaylor One for you I think.

**Comment by ianlancetaylor** [comment]
Yes, this is a bug in https://github.com/spotify/annoy/blob/master/src/annoygomodule.i.  The `argout` typemap allocates memory using `malloc` without any mechanism for freeing that memory.  The Go runtime will never free memory allocated using `malloc`.  This is discussed a bit at https://swig.org/Doc4.0/Go.html#Go_class_memory, although that isn't quite what this code is doing.

This isn't a bug in SWIG and there is nothing to change in the code.  The best we could do here is to improve the documentation.  Any suggestions for what we could write to help people avoid this problem?

**Comment by ojwb** [comment]
From the perspective of someone who doesn't know a lot about Go (certainly nothing useful at this low level), I think the useful thing to provide would be an outline of how to write such a typemap that doesn't leak (the existing docs make me think perhaps by using a finaliser?  If so an example would be helpful), or how to solve the problem another way if there's a better approach (e.g. instead of using `malloc()` can you get ask Go runtime to allocate block the a GC-tracked array from C/C++?  That's effectively how this would work when generating bindings for a typical scripting language).

**Comment by Rikanishu** [comment]
Thanks for pointing the reason of the problem! I've made a forked version without memory leaks (at least, without known ones 😅), and going to make a PR to main repo: 

https://github.com/Rikanishu/annoy-go

I agree with @ojwb that describing the best practices how to deal with Golang slices in SWIG documentation can help preventing such problems with memory management. Maybe advice not to allocate directly to Golang slice in C++ code or advise to wrap vectors or use intermediate type. And especially add an example how code in Golang can get a vector content from C++.
